File: pythonexamples/god_object.py
"""
ANTI-PATTERN: God Object

A class that knows too much or does too much. It has too many responsibilities
and becomes a central point that controls or knows about everything in the system.
Violates the Single Responsibility Principle.
"""

class ApplicationManager:
    """
    GOD OBJECT: This class does EVERYTHING.
    It should be split into separate classes for each responsibility.
    """

    def __init__(self):
        # User management
        self.users = []
        self.current_user = None
        self.user_sessions = {}

        # Database
        self.db_connection = None
        self.db_pool = []

        # Product management
        self.products = []
        self.product_categories = []

        # Order management
        self.orders = []
        self.order_history = []

        # Payment processing
        self.payment_methods = []
        self.transactions = []

        # Email service
        self.email_templates = {}
        self.email_queue = []

        # Logging
        self.logs = []
        self.error_logs = []

        # Configuration
        self.config = {}
        self.feature_flags = {}

        # Cache management
        self.cache = {}
        self.cache_expiry = {}

        # File management
        self.uploaded_files = []
        self.file_storage_path = "/tmp"

        # Authentication
        self.auth_tokens = {}
        self.refresh_tokens = {}

        # Notification system
        self.notifications = []
        self.notification_preferences = {}

        # Analytics
        self.page_views = []
        self.user_events = []

        # Third-party integrations
        self.stripe_api_key = None
        self.aws_credentials = {}
        self.sendgrid_api_key = None

    # USER MANAGEMENT METHODS
    def create_user(self, username, email, password):
        """Should be in UserManager class"""
        pass

    def delete_user(self, user_id):
        """Should be in UserManager class"""
        pass

    def update_user_profile(self, user_id, data):
        """Should be in UserManager class"""
        pass

    def authenticate_user(self, username, password):
        """Should be in AuthenticationService class"""
        pass

    def logout_user(self, user_id):
        """Should be in AuthenticationService class"""
        pass

    def reset_password(self, email):
        """Should be in AuthenticationService class"""
        pass

    # DATABASE METHODS
    def connect_to_database(self):
        """Should be in DatabaseManager class"""
        pass

    def execute_query(self, query):
        """Should be in DatabaseManager class"""
        pass

    def migrate_database(self):
        """Should be in DatabaseManager class"""
        pass

    def backup_database(self):
        """Should be in DatabaseManager class"""
        pass

    # PRODUCT METHODS
    def add_product(self, product):
        """Should be in ProductService class"""
        pass

    def remove_product(self, product_id):
        """Should be in ProductService class"""
        pass

    def update_product_price(self, product_id, new_price):
        """Should be in ProductService class"""
        pass

    def search_products(self, query):
        """Should be in ProductService class"""
        pass

    def get_product_recommendations(self, user_id):
        """Should be in RecommendationEngine class"""
        pass

    # ORDER METHODS
    def create_order(self, user_id, items):
        """Should be in OrderService class"""
        pass

    def cancel_order(self, order_id):
        """Should be in OrderService class"""
        pass

    def calculate_shipping(self, order_id):
        """Should be in ShippingService class"""
        pass

    def track_order(self, order_id):
        """Should be in ShippingService class"""
        pass

    # PAYMENT METHODS
    def process_payment(self, order_id, payment_method):
        """Should be in PaymentService class"""
        pass

    def refund_payment(self, transaction_id):
        """Should be in PaymentService class"""
        pass

    def validate_credit_card(self, card_number):
        """Should be in PaymentService class"""
        pass

    # EMAIL METHODS
    def send_email(self, to, subject, body):
        """Should be in EmailService class"""
        pass

    def send_welcome_email(self, user_id):
        """Should be in EmailService class"""
        pass

    def send_order_confirmation(self, order_id):
        """Should be in EmailService class"""
        pass

    def send_password_reset_email(self, email):
        """Should be in EmailService class"""
        pass

    # LOGGING METHODS
    def log_info(self, message):
        """Should be in Logger class"""
        pass

    def log_error(self, message):
        """Should be in Logger class"""
        pass

    def export_logs(self, format):
        """Should be in Logger class"""
        pass

    # CACHE METHODS
    def cache_set(self, key, value, ttl=3600):
        """Should be in CacheManager class"""
        pass

    def cache_get(self, key):
        """Should be in CacheManager class"""
        pass

    def cache_invalidate(self, key):
        """Should be in CacheManager class"""
        pass

    def cache_clear_all(self):
        """Should be in CacheManager class"""
        pass

    # FILE MANAGEMENT METHODS
    def upload_file(self, file, user_id):
        """Should be in FileStorageService class"""
        pass

    def delete_file(self, file_id):
        """Should be in FileStorageService class"""
        pass

    def get_file_url(self, file_id):
        """Should be in FileStorageService class"""
        pass

    # NOTIFICATION METHODS
    def send_notification(self, user_id, message):
        """Should be in NotificationService class"""
        pass

    def mark_notification_read(self, notification_id):
        """Should be in NotificationService class"""
        pass

    def get_unread_notifications(self, user_id):
        """Should be in NotificationService class"""
        pass

    # ANALYTICS METHODS
    def track_page_view(self, user_id, page):
        """Should be in AnalyticsService class"""
        pass

    def track_event(self, user_id, event_name, properties):
        """Should be in AnalyticsService class"""
        pass

    def generate_analytics_report(self, start_date, end_date):
        """Should be in AnalyticsService class"""
        pass

    # CONFIGURATION METHODS
    def get_config(self, key):
        """Should be in ConfigurationManager class"""
        pass

    def set_config(self, key, value):
        """Should be in ConfigurationManager class"""
        pass

    def is_feature_enabled(self, feature_name):
        """Should be in FeatureFlagManager class"""
        pass

    # INTEGRATION METHODS
    def sync_with_stripe(self):
        """Should be in StripeIntegration class"""
        pass

    def upload_to_s3(self, file):
        """Should be in S3Integration class"""
        pass

    def send_via_sendgrid(self, email_data):
        """Should be in SendGridIntegration class"""
        pass

    # AND MANY MORE METHODS...
    def validate_input(self, data):
        """Should be in ValidationService"""
        pass

    def encrypt_data(self, data):
        """Should be in EncryptionService"""
        pass

    def generate_report(self, report_type):
        """Should be in ReportGenerator"""
        pass

    def schedule_task(self, task):
        """Should be in TaskScheduler"""
        pass

    def check_permissions(self, user_id, resource):
        """Should be in PermissionManager"""
        pass


"""
This God Object violates:
- Single Responsibility Principle (it has dozens of responsibilities)
- Open/Closed Principle (any change requires modifying this massive class)
- Interface Segregation Principle (users are forced to depend on methods they don't use)

It should be refactored into separate classes:
- UserManager
- AuthenticationService
- DatabaseManager
- ProductService
- OrderService
- PaymentService
- EmailService
- Logger
- CacheManager
- FileStorageService
- NotificationService
- AnalyticsService
- ConfigurationManager
- Various integration classes
"""

File: golangexamples/god_object.go
package main

/*
ANTI-PATTERN: God Object

A struct that knows too much or does too much. It has too many responsibilities
and becomes a central point that controls or knows about everything in the system.
Violates the Single Responsibility Principle.
*/

import (
	"fmt"
	"time"
)

// ApplicationManager - GOD OBJECT: This struct does EVERYTHING.
// It should be split into separate types for each responsibility.
type ApplicationManager struct {
	// User management
	users          []User
	currentUser    *User
	userSessions   map[string]*Session
	authTokens     map[string]string
	refreshTokens  map[string]string

	// Database
	dbConnection   interface{}
	dbPool         []interface{}
	dbTransactions []interface{}

	// Product management
	products         []Product
	productCategories []string
	inventory        map[int]int

	// Order management
	orders       []Order
	orderHistory []Order
	orderQueue   []Order

	// Payment processing
	paymentMethods []string
	transactions   []Transaction
	refunds        []Refund

	// Email service
	emailTemplates map[string]string
	emailQueue     []Email
	smtpConfig     map[string]string

	// Logging
	logs       []string
	errorLogs  []string
	auditTrail []string

	// Configuration
	config       map[string]interface{}
	featureFlags map[string]bool
	environment  string

	// Cache management
	cache       map[string]interface{}
	cacheExpiry map[string]time.Time

	// File management
	uploadedFiles    []string
	fileStoragePath  string
	maxFileSize      int64

	// Notification system
	notifications            []Notification
	notificationPreferences  map[int]NotificationPrefs

	// Analytics
	pageViews  []PageView
	userEvents []Event

	// Third-party integrations
	stripeAPIKey    string
	awsCredentials  map[string]string
	sendgridAPIKey  string
}

// Supporting types
type User struct {
	ID       int
	Username string
	Email    string
}

type Session struct {
	ID        string
	UserID    int
	ExpiresAt time.Time
}

type Product struct {
	ID    int
	Name  string
	Price float64
}

type Order struct {
	ID     int
	UserID int
	Items  []int
	Total  float64
}

type Transaction struct {
	ID     string
	Amount float64
	Status string
}

type Refund struct {
	TransactionID string
	Amount        float64
}

type Email struct {
	To      string
	Subject string
	Body    string
}

type Notification struct {
	UserID  int
	Message string
}

type NotificationPrefs struct {
	Email bool
	SMS   bool
	Push  bool
}

type PageView struct {
	UserID int
	Page   string
	Time   time.Time
}

type Event struct {
	UserID int
	Name   string
	Data   map[string]interface{}
}

// NewApplicationManager creates a new god object
func NewApplicationManager() *ApplicationManager {
	return &ApplicationManager{
		users:                   make([]User, 0),
		userSessions:            make(map[string]*Session),
		authTokens:              make(map[string]string),
		refreshTokens:           make(map[string]string),
		products:                make([]Product, 0),
		productCategories:       make([]string, 0),
		inventory:               make(map[int]int),
		orders:                  make([]Order, 0),
		orderHistory:            make([]Order, 0),
		paymentMethods:          make([]string, 0),
		transactions:            make([]Transaction, 0),
		emailTemplates:          make(map[string]string),
		emailQueue:              make([]Email, 0),
		logs:                    make([]string, 0),
		errorLogs:               make([]string, 0),
		config:                  make(map[string]interface{}),
		featureFlags:            make(map[string]bool),
		cache:                   make(map[string]interface{}),
		cacheExpiry:             make(map[string]time.Time),
		notifications:           make([]Notification, 0),
		notificationPreferences: make(map[int]NotificationPrefs),
		pageViews:               make([]PageView, 0),
		userEvents:              make([]Event, 0),
	}
}

// USER MANAGEMENT METHODS - Should be in UserManager
func (am *ApplicationManager) CreateUser(username, email, password string) (*User, error) {
	// Should be in UserManager struct
	return nil, nil
}

func (am *ApplicationManager) DeleteUser(userID int) error {
	// Should be in UserManager struct
	return nil
}

func (am *ApplicationManager) UpdateUserProfile(userID int, data map[string]interface{}) error {
	// Should be in UserManager struct
	return nil
}

func (am *ApplicationManager) GetUserByID(userID int) (*User, error) {
	// Should be in UserManager struct
	return nil, nil
}

// AUTHENTICATION METHODS - Should be in AuthenticationService
func (am *ApplicationManager) AuthenticateUser(username, password string) (string, error) {
	// Should be in AuthenticationService struct
	return "", nil
}

func (am *ApplicationManager) LogoutUser(userID int) error {
	// Should be in AuthenticationService struct
	return nil
}

func (am *ApplicationManager) ResetPassword(email string) error {
	// Should be in AuthenticationService struct
	return nil
}

func (am *ApplicationManager) ValidateToken(token string) (bool, error) {
	// Should be in AuthenticationService struct
	return false, nil
}

func (am *ApplicationManager) RefreshAuthToken(refreshToken string) (string, error) {
	// Should be in AuthenticationService struct
	return "", nil
}

// DATABASE METHODS - Should be in DatabaseManager
func (am *ApplicationManager) ConnectToDatabase() error {
	// Should be in DatabaseManager struct
	return nil
}

func (am *ApplicationManager) ExecuteQuery(query string, args ...interface{}) (interface{}, error) {
	// Should be in DatabaseManager struct
	return nil, nil
}

func (am *ApplicationManager) MigrateDatabase() error {
	// Should be in DatabaseManager struct
	return nil
}

func (am *ApplicationManager) BackupDatabase() error {
	// Should be in DatabaseManager struct
	return nil
}

func (am *ApplicationManager) RollbackTransaction() error {
	// Should be in DatabaseManager struct
	return nil
}

// PRODUCT METHODS - Should be in ProductService
func (am *ApplicationManager) AddProduct(product Product) error {
	// Should be in ProductService struct
	return nil
}

func (am *ApplicationManager) RemoveProduct(productID int) error {
	// Should be in ProductService struct
	return nil
}

func (am *ApplicationManager) UpdateProductPrice(productID int, newPrice float64) error {
	// Should be in ProductService struct
	return nil
}

func (am *ApplicationManager) SearchProducts(query string) ([]Product, error) {
	// Should be in ProductService struct
	return nil, nil
}

func (am *ApplicationManager) GetProductRecommendations(userID int) ([]Product, error) {
	// Should be in RecommendationEngine struct
	return nil, nil
}

func (am *ApplicationManager) UpdateInventory(productID, quantity int) error {
	// Should be in InventoryService struct
	return nil
}

// ORDER METHODS - Should be in OrderService
func (am *ApplicationManager) CreateOrder(userID int, items []int) (*Order, error) {
	// Should be in OrderService struct
	return nil, nil
}

func (am *ApplicationManager) CancelOrder(orderID int) error {
	// Should be in OrderService struct
	return nil
}

func (am *ApplicationManager) GetOrderStatus(orderID int) (string, error) {
	// Should be in OrderService struct
	return "", nil
}

func (am *ApplicationManager) CalculateShipping(orderID int) (float64, error) {
	// Should be in ShippingService struct
	return 0, nil
}

func (am *ApplicationManager) TrackOrder(orderID int) (string, error) {
	// Should be in ShippingService struct
	return "", nil
}

// PAYMENT METHODS - Should be in PaymentService
func (am *ApplicationManager) ProcessPayment(orderID int, paymentMethod string) error {
	// Should be in PaymentService struct
	return nil
}

func (am *ApplicationManager) RefundPayment(transactionID string) error {
	// Should be in PaymentService struct
	return nil
}

func (am *ApplicationManager) ValidateCreditCard(cardNumber string) (bool, error) {
	// Should be in PaymentService struct
	return false, nil
}

func (am *ApplicationManager) GetPaymentHistory(userID int) ([]Transaction, error) {
	// Should be in PaymentService struct
	return nil, nil
}

// EMAIL METHODS - Should be in EmailService
func (am *ApplicationManager) SendEmail(to, subject, body string) error {
	// Should be in EmailService struct
	return nil
}

func (am *ApplicationManager) SendWelcomeEmail(userID int) error {
	// Should be in EmailService struct
	return nil
}

func (am *ApplicationManager) SendOrderConfirmation(orderID int) error {
	// Should be in EmailService struct
	return nil
}

func (am *ApplicationManager) SendPasswordResetEmail(email string) error {
	// Should be in EmailService struct
	return nil
}

func (am *ApplicationManager) QueueEmail(email Email) error {
	// Should be in EmailService struct
	return nil
}

// LOGGING METHODS - Should be in Logger
func (am *ApplicationManager) LogInfo(message string) {
	// Should be in Logger struct
	am.logs = append(am.logs, fmt.Sprintf("[INFO] %s", message))
}

func (am *ApplicationManager) LogError(message string) {
	// Should be in Logger struct
	am.errorLogs = append(am.errorLogs, fmt.Sprintf("[ERROR] %s", message))
}

func (am *ApplicationManager) ExportLogs(format string) (string, error) {
	// Should be in Logger struct
	return "", nil
}

func (am *ApplicationManager) ClearLogs() {
	// Should be in Logger struct
	am.logs = make([]string, 0)
}

// CACHE METHODS - Should be in CacheManager
func (am *ApplicationManager) CacheSet(key string, value interface{}, ttl time.Duration) {
	// Should be in CacheManager struct
	am.cache[key] = value
	am.cacheExpiry[key] = time.Now().Add(ttl)
}

func (am *ApplicationManager) CacheGet(key string) (interface{}, bool) {
	// Should be in CacheManager struct
	value, exists := am.cache[key]
	return value, exists
}

func (am *ApplicationManager) CacheInvalidate(key string) {
	// Should be in CacheManager struct
	delete(am.cache, key)
}

func (am *ApplicationManager) CacheClearAll() {
	// Should be in CacheManager struct
	am.cache = make(map[string]interface{})
}

// FILE MANAGEMENT METHODS - Should be in FileStorageService
func (am *ApplicationManager) UploadFile(file []byte, userID int) (string, error) {
	// Should be in FileStorageService struct
	return "", nil
}

func (am *ApplicationManager) DeleteFile(fileID string) error {
	// Should be in FileStorageService struct
	return nil
}

func (am *ApplicationManager) GetFileURL(fileID string) (string, error) {
	// Should be in FileStorageService struct
	return "", nil
}

// NOTIFICATION METHODS - Should be in NotificationService
func (am *ApplicationManager) SendNotification(userID int, message string) error {
	// Should be in NotificationService struct
	return nil
}

func (am *ApplicationManager) MarkNotificationRead(notificationID int) error {
	// Should be in NotificationService struct
	return nil
}

func (am *ApplicationManager) GetUnreadNotifications(userID int) ([]Notification, error) {
	// Should be in NotificationService struct
	return nil, nil
}

// ANALYTICS METHODS - Should be in AnalyticsService
func (am *ApplicationManager) TrackPageView(userID int, page string) {
	// Should be in AnalyticsService struct
	am.pageViews = append(am.pageViews, PageView{UserID: userID, Page: page, Time: time.Now()})
}

func (am *ApplicationManager) TrackEvent(userID int, eventName string, properties map[string]interface{}) {
	// Should be in AnalyticsService struct
	am.userEvents = append(am.userEvents, Event{UserID: userID, Name: eventName, Data: properties})
}

func (am *ApplicationManager) GenerateAnalyticsReport(startDate, endDate time.Time) (string, error) {
	// Should be in AnalyticsService struct
	return "", nil
}

// CONFIGURATION METHODS - Should be in ConfigurationManager
func (am *ApplicationManager) GetConfig(key string) (interface{}, bool) {
	// Should be in ConfigurationManager struct
	value, exists := am.config[key]
	return value, exists
}

func (am *ApplicationManager) SetConfig(key string, value interface{}) {
	// Should be in ConfigurationManager struct
	am.config[key] = value
}

func (am *ApplicationManager) IsFeatureEnabled(featureName string) bool {
	// Should be in FeatureFlagManager struct
	return am.featureFlags[featureName]
}

func main() {
	/*
		This God Object violates:
		- Single Responsibility Principle (it has dozens of responsibilities)
		- Open/Closed Principle (any change requires modifying this massive struct)
		- Interface Segregation Principle (users depend on methods they don't use)

		It should be refactored into separate structs:
		- UserManager
		- AuthenticationService
		- DatabaseManager
		- ProductService
		- OrderService
		- PaymentService
		- EmailService
		- Logger
		- CacheManager
		- FileStorageService
		- NotificationService
		- AnalyticsService
		- ConfigurationManager
		- Various integration services
	*/

	am := NewApplicationManager()
	fmt.Printf("God object created with %d responsibilities\n", 15)
	_ = am
}
